### PR TITLE
Wgan implementation + Config modifications

### PR DESCRIPTION
**Pull request by Akssi**
model config contains model args and is passed as a whole to the model factory

### CODE REVIEW

File: configs/gan.yaml [hunk 1]
```diff
@@ -1 +1,12 @@
-# TODO: create GAN config
\ No newline at end of file
+# TODO: create GAN config
```
Comment (Olivier-tl) [inline configs/gan.yaml:1]: We can remove

File: models/gans/wgan.py [hunk 2]
```diff
@@ -1,12 +1,181 @@
 # TODO: Implement the wgan architecture (issue #6)
```
Comment (Olivier-tl) [inline models/gans/wgan.py:1]: We can remove

File: models/gans/wgan.py [hunk 3]
```diff
@@ -1,12 +1,181 @@
...
+import torch.optim as optim
+from torch.autograd import Variable
+from torch import autograd
+import time as t
+import matplotlib.pyplot as plt
+plt.switch_backend('agg')
+import os
+from itertools import chain
+from torchvision import utils as utils
+from utils.logging import logging
```
Comment (Olivier-tl) [inline models/gans/wgan.py:14]: I think we should follow PEP8 for the imports! (default package -> third party -> custom) separated with a space

Comment (Olivier-tl) [reply models/gans/wgan.py:14]: import os
import time as t
from itertools import chain

import torch
import torch.nn as nn
import torch.optim as optim
from torch.autograd import Variable
from torch import autograd
from torchvision import utils as utils
import matplotlib.pyplot as plt
plt.switch_backend('agg')

from utils import logging

Comment (Olivier-tl) [reply models/gans/wgan.py:14]: There is some dependencies that are not used like matplotlib

File: models/gans/wgan.py [hunk 4]
```diff
@@ -1,12 +1,181 @@
...
-class Discriminator:
-    pass
\ No newline at end of file
+        # WGAN values from paper
+        self.learning_rate = 1e-4
+        self.b1 = 0.5
+        self.b2 = 0.999
+        self.batch_size = 64
+
+        # WGAN_gradient penalty uses ADAM
```
Comment (Olivier-tl) [inline models/gans/wgan.py:34]: Should this be in the GANTrainer instead?

File: utils/misc.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+
+class AttrDict(dict):
```
Comment (Olivier-tl) [inline utils/misc.py:2]: That AttrDict thing is amazing! Can we trash the whole Config class and use this instead? 👀

File: models/gans/wgan.py [hunk 6]
```diff
@@ -1,12 +1,181 @@
...
+
+    def load_model(self, D_model_filename, G_model_filename):
+        D_model_path = os.path.join(os.getcwd(), D_model_filename)
+        G_model_path = os.path.join(os.getcwd(), G_model_filename)
+        self.D.load_state_dict(torch.load(D_model_path))
+        self.G.load_state_dict(torch.load(G_model_path))
+        print('Generator model loaded from {}.'.format(G_model_path))
+        print('Discriminator model loaded from {}-'.format(D_model_path))
+
+    def generate_latent_walk(self, number):
```
Comment (Olivier-tl) [inline models/gans/wgan.py:80]: We don't need that but it can be cool to keep.

File: models/gans/wgan.py [hunk 7]
```diff
@@ -1,12 +1,181 @@
...
+import matplotlib.pyplot as plt
+plt.switch_backend('agg')
+import os
+from itertools import chain
+from torchvision import utils as utils
+from utils.logging import logging
 
-class WGAN:
-    pass
+SAVE_PER_TIMES = 100
```
Comment (Olivier-tl) [inline models/gans/wgan.py:15]: I think we should put the ref of the code on the first line: 
https://github.com/Zeleni9/pytorch-wgan/blob/master/models/wgan_gradient_penalty.py

**Review by Olivier-tl — APPROVED** [review]

File: models/gans/wgan.py (hunk 4 above)
Comment (wwwnev) [reply models/gans/wgan.py:34]: I thing that if the trainer classes are initialized w/ DataLoaders, the batch_size would have to be set before initializing those classes.

**Review by wwwnev — APPROVED** [review]

File: models/gans/wgan.py (hunk 4 above)
Comment (Akssi) [reply models/gans/wgan.py:34]: I left a bunch of code that could be transferred into the trainer

File: utils/misc.py (hunk 5 above)
Comment (Akssi) [reply utils/misc.py:2]: Maybe depends if we have to add other methods to config

**Pull request merged**